src/core/cgroup.c

```diff
@@ -2543,6 +2562,12 @@ unsigned manager_dispatch_cgroup_realize_queue(Manager *m) {
         assert(m);
 
         state = manager_state(m);
+        if (state == MANAGER_STOPPING && !m->shutdown_units_refreshed) {
+                /* Units which are refreshed after startup are also refreshed
+                 * at shutdown time. */
+                manager_invalidate_startup_units(m);
+                m->shutdown_units_refreshed = true;
+        }
```

Comment:
this doesn't look like the right place to insert this.

if you look into manager_state() you'll find that MANAGER_STOPPING is returned whenever the special unit "shutdown.target" is enqueued or active. Thus, it's important to invalidate the startup unit cgroups stuff whenever that unit is enqueued. i.e. add a call to manager_invalidate_startup_units() to manager_add_job() if the right unit is enqueued.

Reply:
I don't think SPECIAL_SHUTDOWN_TARGET is directly enqueued as a job? I see SPECIAL_{HALT,POWEROFF,REBOOT,KEXEC}_TARGET be enqueued by not SPECIAL_SHUTDOWN_TARGET. The later seems to be installed indirectly.

Thus I don't think I can add a general catch all in manager_add_job() which would call manager_invalidate_startup_units() when any of the HALT/POWEROFF/REBOOT/KEXEC jobs are added. I could write some code to catch all these targets but that doesn't seem very nice.

I've pushed changes which do work, manager_invalidate_startup_units() is called via job_install(). This doesn't feel great either, so looking for your feedback again :)

Reply:
@poettering would you be able to take a look when you have a moment please?  Thanks!